services/wallet/history/service.go

```diff
@@ -0,0 +1,339 @@
+package history
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+
+	statustypes "github.com/status-im/status-go/eth-node/types"
+	"github.com/status-im/status-go/multiaccounts/accounts"
+	statusrpc "github.com/status-im/status-go/rpc"
+	"github.com/status-im/status-go/rpc/network"
+
+	"github.com/status-im/status-go/services/wallet/chain"
+	"github.com/status-im/status-go/services/wallet/token"
+	"github.com/status-im/status-go/services/wallet/walletevent"
+)
+
+// EventBalanceHistoryUpdateStarted and EventBalanceHistoryUpdateDone are used to notify the UI that balance history is being updated
+const (
+	EventBalanceHistoryUpdateStarted           walletevent.EventType = "wallet-balance-history-update-started"
+	EventBalanceHistoryUpdateFinished          walletevent.EventType = "wallet-balance-history-update-finished"
+	EventBalanceHistoryUpdateFinishedWithError walletevent.EventType = "wallet-balance-history-update-finished-with-error"
+
+	balanceHistoryUpdateInterval = 12 * time.Hour
+)
+
+type Service struct {
+	balance        *Balance
+	db             *sql.DB
+	eventFeed      *event.Feed
+	rpcClient      *statusrpc.Client
+	networkManager *network.Manager
+	tokenManager   *token.Manager
+	serviceContext context.Context
+	cancelFn       context.CancelFunc
+
+	timer          *time.Timer
+	visibleSymbols []string
+	mutex          sync.Mutex // Protects access to visibleSymbols
+}
+
+func NewService(db *sql.DB, eventFeed *event.Feed, rpcClient *statusrpc.Client, tokenManager *token.Manager) *Service {
+	return &Service{
+		balance:        NewBalance(NewBalanceDB(db)),
+		db:             db,
+		eventFeed:      eventFeed,
+		rpcClient:      rpcClient,
+		networkManager: rpcClient.NetworkManager,
+		tokenManager:   tokenManager,
+	}
+}
+
+func (s *Service) Stop() {
+	s.cancelFn()
+}
+
+func (s *Service) triggerEvent(eventType walletevent.EventType, account statustypes.Address, message string) {
+	s.eventFeed.Send(walletevent.Event{
+		Type: eventType,
+		Accounts: []common.Address{
+			common.Address(account),
```

Comment:
isn't account always empty?

Reply:
No, only in case of cancellation or critical error. Otherwise, it will contain the account address when starting and finishing the fetching for the specific address. Plan to use it in UI when visualizing an account's graph to update when data is available.